src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,383 @@
+  * `gen` type.  However, it also has an `indices` argument.  This indicates the
+  * indices at which the [[SparseVec]] is allowed to have data.  Additionally,
+  * the behavior of a [[SparseVec]] around what happens if a value is read from
+  * a value not in the `indices`:
+  *
+  * - `defaultValue` sets the default value that is read from an index between
+  *   the zeroth index and the largest value in `indices` that is not in the
+  *   `indices`.
+  * - `outOfBoundsValue` sets the behavior when reading a value larger than the
+  *   largest value in `indices`.
+  *
+  * The reason for this configurability is to enable exact compatibility with an
+  * equivalent dense [[Vec]] of the same size and initialized to a given value.
+  * Specifically, use [[SparseVec.DefaultValueBehavior.DynamicIndexEquivalent]]
+  * and [[SparseVec.OutOfBoundsBehavior.First]] to make this behave as such:
+  *
+  * 1. The [[SparseVec]] has a default value of how a FIRRTL compiler compiles
+  *    [[DontCare]] for a dynamic index.
+  *
+  * 2. The [[SparseVec]] out-of-bounds behavior returns the zeroth element if a
+  *    zeroth element.  Otherwise, it returns a [[DontCare]].
+  *
+  * Note that this [[DontCare]] is likely not a true "don't care" that will be
+  * optimized to any value.  Instead, it is a value equal to how a FIRRTL
+  * compiler chooses to optimize a dynamic index into a wire vector initialized
+  * with a [[DontCare]].  This has historically been zero.
+  *
+  * Once created, a [[SparseVec]] can be written or read from as a [[Record]].
+  * It may also be read from using a dynamic index, but not written to.  Neither
+  * the default value nor the out-of-bounds value may be written to.  The
+  * dynamic index type is conifgurable and may be one of:
+  *
+  * - [[SparseVec.Lookup.Binary]] to convert the [[SparseVec]] index into a
+  *   binary index into a dense vector.
+  * - [[SparseVec.Lookup.OneHot]] to convert the [[SparseVec]] index into a
+  *   one-hot encoded index into a dense vector using [[Mux1H]].
+  * - [[SparseVec.Lookup.IfElse]] to use a sequence of [[when]] statements.
+  *
+  * A [[SparseVec]] will take up storage equal to the size of the provided
+  * mapping argument with one additional slot for the default value, if one is
+  * needed.
+  *
+  * @param size the apparent size of the vector
+  * @param gen the element type of the vector
+  * @param indices the indices of the vector which are valid
+  * @param defaultValue the default value behavior when accessing an index not
+  * in the `indices`
+  * @param outOfBoundsValue the out-of-bounds behavior when accessing an index
+  * larger than the largest value in `indices`
+  */
+class SparseVec[A <: Data](
+  size:             Int,
+  gen:              => A,
+  indices:          Seq[Int],
+  defaultValue:     DefaultValueBehavior.Type = DefaultValueBehavior.DynamicIndexEquivalent,
+  outOfBoundsValue: OutOfBoundsBehavior.Type = OutOfBoundsBehavior.First)
```

Comment:
Do we want the default behavior to be the backwards-compatible [with sparse vec implemented using a Vec with DontCare holes] or do we want this new thing to take advantage of undefined behavior by default while still letting the user opt-in to the backwards compatible behavior?

I'm asking because I think we really could go either way here.

Reply:
Probably new behavior.

Reply:
I'll change it to that.